src/neo/IO/MemoryReader.cs

```diff
@@ -0,0 +1,238 @@
+// Copyright (C) 2015-2022 The Neo Project.
+// 
+// The neo is free software distributed under the MIT software license, 
+// see the accompanying file LICENSE in the main directory of the
+// project or http://www.opensource.org/licenses/mit-license.php 
+// for more details.
+// 
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using System;
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+
+namespace Neo.IO
+{
+    public ref struct MemoryReader
+    {
+        private readonly ReadOnlyMemory<byte> memory;
+        private readonly ReadOnlySpan<byte> span;
+        private int pos = 0;
+
+        public int Position => pos;
+
+        public MemoryReader(ReadOnlyMemory<byte> memory)
+        {
+            this.memory = memory;
+            this.span = memory.Span;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsurePosition(int move)
+        {
+            if (pos + move > span.Length) throw new FormatException();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public byte Peek()
+        {
+            EnsurePosition(1);
+            return span[pos];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool ReadBoolean()
+        {
+            return ReadByte() switch
+            {
+                0 => false,
+                1 => true,
```

Comment:
I think that in the previous version, `true` it's something different than 0. it could be?

Reply:
Yes. But I think we need a deterministic value.